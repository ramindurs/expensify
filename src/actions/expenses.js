import {v4 as uuidv4} from "uuid";
import database from '../firebase/firebase';

const FB_REF = 'expenses';

export const addExpense = (expense) => ({
        type: 'ADD_EXPENSE',
        expense
    }
);

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = {
            description,
            note,
            amount,
            createdAt
        };
        return database().ref(FB_REF).push(expense).then((fbRef) => {
            dispatch(addExpense({
                id: fbRef.key,
                ...expense
            }))
        })
    };
};

export const removeExpense = (id) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = (id) => {
    return (dispatch) => {
        const ref = `${FB_REF}/${id}`;
        return database().ref(ref).remove().then(() => {
            return dispatch(removeExpense(id));
        }).catch((error) => {
            console.error('Error removing ', ref);
        });
    };
};

export const editExpense = ({id, updates}) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = ({id, updates}) => {
    return (dispatch) => {
        const ref = `${FB_REF}/${id}`;
        return database().ref(ref).update(updates).then(() => {
            return dispatch(editExpense({id, updates}));
        }).catch((error) => {
            console.error('Error updating', ref);
        });
    }
};

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        return database().ref(FB_REF).once('value')
            .then((dataSnapshot) => {
                const expenses = [];
                dataSnapshot.forEach((childSnapshot) => {
                    expenses.push({
                        ...childSnapshot.val(),
                        id: childSnapshot.key
                    });
                });
                return dispatch(setExpenses(expenses));
            });
    };
};
