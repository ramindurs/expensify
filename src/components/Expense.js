import React from 'react';
import {Link} from 'react-router-dom';

export const Expense = ({ id, description, amount, createdAt, dispatch, history}) => {
    return (
        <div>
            <p>{id} <Link to={`/edit/${id}` }>{description}</Link> {amount} {createdAt}</p>
        </div>
    );
};

export default Expense;
