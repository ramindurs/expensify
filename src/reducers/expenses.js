
const expensesRedecuerDefaultState = [];

export default (state = expensesRedecuerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            return state.filter(ex => ex.id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map((expense) => expense.id !== action.id ? expense : {
                ...expense,
                ...action.updates
            });
        case 'SET_EXPENSES':
            return action.expenses;
        default:
            return state;
    }
};