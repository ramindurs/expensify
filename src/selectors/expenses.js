import moment from "moment";

export default (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment) : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment) : true;
        const textMatch = expense.description.toUpperCase().indexOf(text.toUpperCase()) !== -1;
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => (sortBy === 'date' ? b.createdAt - a.createdAt: b.amount - a.amount));
};