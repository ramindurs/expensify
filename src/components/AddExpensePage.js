import React from 'react';
import {connect} from 'react-redux';
import {addExpense} from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.onSubmit(expense);
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <h3>Add Expenses</h3>
                <ExpenseForm
                    submitMessage={'Add Expense'}
                    onSubmit={this.onSubmit}/>
            </div>
        );
    }
}

const mapDispatchTpProps = (dispatch) => {
    return {
        onSubmit: (expense) => dispatch(addExpense(expense))
    };
};

export default connect(undefined, mapDispatchTpProps)(AddExpensePage);