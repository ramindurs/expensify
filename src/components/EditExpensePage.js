import React from 'react';
import {connect} from "react-redux";
import ExpenseForm from "./ExpenseForm";
import {editExpense, removeExpense} from "../actions/expenses";

export class EditExpensePage extends React.Component{
    removeExpense = (id) => {
        this.props.removeExpense(id);
        this.props.history.push('/');
    };
    editExpense = (id, expense) => {
        this.props.editExpense(id, expense);
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <h3>Edit Expense ID {this.props.expense.id}</h3>
                <ExpenseForm
                    expenseToEdit={this.props.expense}
                    submitMessage={'Update Expense'}
                    onSubmit={(expense) => {this.editExpense(this.props.expense.id, expense)}}
                />
                <button onClick={() => {this.removeExpense(this.props.expense.id)}
                }>Delete</button>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find(e => e.id === props.match.params.id)
    };
};

const matchDispatchToProps = (dispatch, props) => {
    return {
        editExpense: (id, expense) => dispatch(editExpense({
            id: id,
            updates: expense
        })),
        removeExpense: (id) => dispatch(removeExpense({id}))
    };
};

export default connect(mapStateToProps, matchDispatchToProps)(EditExpensePage);