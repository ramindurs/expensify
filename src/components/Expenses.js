import React from 'react';
import {connect} from "react-redux";
import Expense from "./Expense";
import selectExpenses from '../selectors/expenses';

export const Expenses = (props) => (
    <div>
        <h1>Expenses List</h1>
        {props.expenses.map((expense, index) =>
            <Expense key={index}
                     {...expense}
            />)}
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
};

export default connect(mapStateToProps)(Expenses);
