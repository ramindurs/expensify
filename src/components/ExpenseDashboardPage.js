import React from 'react';
import Expenses from "./Expenses";
import ExpensesFilter from "./ExpensesFilter";

const ExpenseDashboardPage = () => (
    <div>
        <ExpensesFilter/>
        <Expenses/>
    </div>
);

export default ExpenseDashboardPage;