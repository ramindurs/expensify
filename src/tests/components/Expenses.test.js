import React from 'react';
import {shallow} from 'enzyme';
import {Expenses} from "../../components/Expenses";
import expenses from "../fixtures/expenses";

test('should render expense list with expenses', () => {
    const wrapper = shallow(<Expenses expenses={expenses}/>);

    expect(wrapper).toMatchSnapshot();
});

test('should render expense list with no expeses', () => {
    const wrapper = shallow(<Expenses expenses={[]}/>);

    expect(wrapper).toMatchSnapshot();
});