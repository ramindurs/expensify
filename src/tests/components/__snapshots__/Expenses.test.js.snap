// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list with expenses 1`] = `
<div>
  <h1>
    Expenses List
  </h1>
  <Expense
    amount={1500}
    createdAt={1000}
    description="Water Bill"
    id={1}
    key="0"
    note=""
  />
  <Expense
    amount={20000}
    createdAt={2000}
    description="Electric bill"
    id={2}
    key="1"
    note=""
  />
  <Expense
    amount={3000}
    createdAt={3000}
    description="Telephone"
    id={3}
    key="2"
    note=""
  />
</div>
`;

exports[`should render expense list with no expeses 1`] = `
<div>
  <h1>
    Expenses List
  </h1>
</div>
`;
